### DISCUSSION TITLE
CBC Multithreading Time Limit

### DISCUSSION THREAD

**Discussion by ebongo1** [coin-or/pulp#486] [Q&A]
I have been reading through a few online discussions on GitHub and other sources ([https://groups.google.com/g/pulp-or-discuss/c/IKE--kT6c5U/m/Y0OUJCqoBQAJ](url), [https://github.com/coin-or/pulp/issues/463](url)) about using multiple threads with CBC, but I haven’t been able to find a direct answer to my question. So I wanted to reach out to you directly about it. Does the time limit correlate to the number of threads passed in to the solver? So for instance when I set “threads=4” and “time limit=3600”, does that mean each thread gets a time limit of 3600/4 seconds? I.e. does CBC use CPU time or wall time for the time limit?

**Comment by pchtsp** [comment]
As you suspect, CBC uses CPU time for the time limit. So, it will indeed use more or less 3600/4 seconds for each thread (more or less because for the root relaxation it uses only one thread).

Maybe @tkralphs can complement the answer here. In fact, this question could be also moved / copied to the cbc github here: https://github.com/coin-or/Cbc/

**Reply by tkralphs** [reply]
This could be a question better asked in the Cbc forum except for the fact that this is something that can be controlled by parameter. So if the default of `cpu` is not what you want (and it probably isn't what you want), you can change it within PuLP when you call Cbc so there is more uniformity across solvers. Just add `-timeMode elapsed` to the Cbc command-line. I'm not too sure if users can do this themselves or not.

**Reply by pchtsp** [reply]
ah! Good to know that, thanks @tkralphs. I think we will probably set this configuration by default in CBC from PuLP, so it matches the defaults for GUROBI and CPLEX. In the meantime and as a workaround, it is indeed possible to pass arbitrary parameters to CBC from PuLP so this should work right now:

```python

import pulp
solver = pulp.PULP_CBC_CMD(timeLimit=3600, threads=4, options=["timeMode elapsed"])

```
(I did a quick test in my current model and it apparently passed the parameter correctly to CBC).

**Reply by ebongo1** [reply]
Oh ok that is very helpful to know! I still will move my original question over to the CBC github, but thank you both!

**Comment by pchtsp** [comment]
Thanks to @tkralphs comment, I've added the `timeMode="elapsed"` by default for CBC in PuLP. So now, if you re-install pulp (v2.5.1) `timeLimit` is measured in walltime by default:

```python
import pulp
# this counts wall-time and stops after 3600 seconds.
solver = pulp.PULP_CBC_CMD(timeLimit=3600, threads=4)
```

You can change that to cpu time by doing the following:

```python
import pulp
# this sums the time in each cpu, so stops roughly after 3600/4 seconds, as before the change
solver = pulp.PULP_CBC_CMD(timeLimit=3600, threads=4, timeMode='cpu')
```
